app/models/product_for_cart.rb

```diff
@@ -0,0 +1,118 @@
+class ProductForCart
+
+  include TextHelpers::Translation
+
+  attr_accessor :error_message
+  attr_accessor :error_path
+
+  def initialize(product)
+    @product = product
+  end
+
+  def purchasable_by?(acting_user, session_user)
+    error = checks(acting_user, session_user).detect do |check|
+      result = check.call
+      [error_path, error_message].any?(&:present?) || result == false
```

Comment:
I think that I would prefer if each of the `?` methods  (or their procs) returned a proper boolean as the `?` convention usually indicates. Then we wouldn't need the checks for the attr_readers.